### PR TITLE
Bug 1979038: oVirt: include master IPs when running must gather

### PR DESCRIPTION
**Pull request by eslutsky**
`gather bootstrap` command will collect logs from the VMs only if the boostrap VM still exists.
this change adds the control planes VMs IP information, resulting the master logs
to be collected as well.

Signed-off-by: Evgeny Slutsky <eslutsky@redhat.com>

### CODE REVIEW

File: pkg/terraform/gather/ovirt/ip.go [hunk 1]
```diff
@@ -55,9 +55,16 @@ func getReportedDevices(c *ovirtsdk4.Connection, vmID string) (*ovirtsdk4.Report
 	return reportedDeviceSlice, nil
 }
 
-func findVirtualMachineIP(c *ovirtsdk4.Connection, moRefValue string) (string, error) {
+func findVirtualMachineIP(moRefValue string) (string, error) {
 
-	reportedDeviceSlice, err := getReportedDevices(c, moRefValue)
+	client, err := ovirt.NewConnection()
```
Comment (janoszen) [inline pkg/terraform/gather/ovirt/ip.go:60]: Why are we creating a new oVirt connection here?

File: pkg/terraform/gather/ovirt/ip.go [hunk 2]
```diff
@@ -69,52 +76,54 @@ func findVirtualMachineIP(c *ovirtsdk4.Connection, moRefValue string) (string, e
 				ipres, hasAddress := ip.Address()
 				if hasAddress {
 					if checkPortIsOpen(ipres, bootstrapSSHPort) {
-						logrus.Debugf("ovirt vm id: %s , found usable bootstrap IP %s", moRefValue, ipres)
+						logrus.Debugf("ovirt vm id: %s , found usable  IP Address: %s", moRefValue, ipres)
```
Comment (janoszen) [inline pkg/terraform/gather/ovirt/ip.go:79]: Is this log useful to the user anyhow or is this just debug code?

File: pkg/terraform/gather/ovirt/ip.go [hunk 3]
```diff
@@ -55,9 +55,16 @@ func getReportedDevices(c *ovirtsdk4.Connection, vmID string) (*ovirtsdk4.Report
...
+	client, err := ovirt.NewConnection()
+	if err != nil {
+		return "", fmt.Errorf("failed to initialize connection to ovirt-engine's %s", err)
+	}
+	defer client.Close()
+
+
+	reportedDeviceSlice, err := getReportedDevices(client, moRefValue)
 	if err != nil {
 		return "", err
```
Comment (janoszen) [inline pkg/terraform/gather/ovirt/ip.go:69]: This error should possibly be wrapped so we have context.

File: pkg/terraform/gather/ovirt/ip.go [hunk 4]
```diff
@@ -69,52 +76,54 @@ func findVirtualMachineIP(c *ovirtsdk4.Connection, moRefValue string) (string, e
...
-
-	br, err := terraform.LookupResource(tfs, "module.bootstrap", "ovirt_vm", "bootstrap")
-	if err != nil {
-		return "", errors.Wrap(err, "failed to lookup bootstrap")
+		return []string{""}, errors.Wrapf(err, "failed to lookup %s VM",name)
 	}
 
 	if len(br.Instances) == 0 {
-		return "", errors.New("no bootstrap instance found")
+		return []string{""}, errors.New(fmt.Sprintf("no %s instance found",name))
```
Comment (janoszen) [inline pkg/terraform/gather/ovirt/ip.go]: Returning an empty IP seems like a workaround for something. It seems that this should not be handled here. If yes, we should add comments explaining what this does and why it is ok to return an empty string IP.

File: pkg/terraform/gather/ovirt/ip.go [hunk 5]
```diff
@@ -69,52 +76,54 @@ func findVirtualMachineIP(c *ovirtsdk4.Connection, moRefValue string) (string, e
...
 	}
-	return ip, nil
+
+	return ips, nil
+}
+
+// BootstrapIP returns the ip address for bootstrap host.
+func BootstrapIP(tfs *terraform.State) (string, error) {
+	ips, err := lookupVMResources(tfs, "module.bootstrap", "bootstrap")
+	return ips[0], err
```
Comment (janoszen) [inline pkg/terraform/gather/ovirt/ip.go:123]: Add error handling here if ips has no entries.

File: pkg/terraform/gather/ovirt/ip.go [hunk 6]
```diff
@@ -69,52 +76,54 @@ func findVirtualMachineIP(c *ovirtsdk4.Connection, moRefValue string) (string, e
...
+		vmid, found, err := unstructured.NestedString(instance.Attributes, "id")
+		if err != nil {
+			return []string{""}, errors.Wrapf(err, "failed to lookup %s managed object reference",name)
+		}
+		if !found {
+			return []string{""}, errors.Errorf("failed to lookup %s by vmID: %s", name, vmid)
+		}
+		ip, err := findVirtualMachineIP(vmid)
+		if err != nil {
+			return []string{""}, errors.Wrapf(err, "failed to lookup %s ipv4 address",name)
```
Comment (janoszen) [inline pkg/terraform/gather/ovirt/ip.go]: Same as above.

**Review by janoszen — CHANGES_REQUESTED** [review]

File: pkg/terraform/gather/ovirt/ip.go (hunk 2 above)
Comment (eslutsky) [reply pkg/terraform/gather/ovirt/ip.go:79]: its for debug

File: pkg/terraform/gather/ovirt/ip.go (hunk 1 above)
Comment (eslutsky) [reply pkg/terraform/gather/ovirt/ip.go]: moved to the connection one level up (outside the for loop)

File: pkg/terraform/gather/ovirt/ip.go (hunk 4 above)
Comment (eslutsky) [reply pkg/terraform/gather/ovirt/ip.go]: okay , in case of error we can return nil instead

File: pkg/terraform/gather/ovirt/ip.go [hunk 7]
```diff
@@ -55,9 +55,8 @@ func getReportedDevices(c *ovirtsdk4.Connection, vmID string) (*ovirtsdk4.Report
 	return reportedDeviceSlice, nil
 }
 
-func findVirtualMachineIP(c *ovirtsdk4.Connection, moRefValue string) (string, error) {
-
-	reportedDeviceSlice, err := getReportedDevices(c, moRefValue)
+func findVirtualMachineIP(moRefValue string,client *ovirtsdk4.Connection) (string, error) {
```
Comment (janoszen) [inline pkg/terraform/gather/ovirt/ip.go:58]: This change seems unnecessary. Not a big deal, but still.

File: pkg/terraform/gather/ovirt/ip.go (hunk 5 above)
Comment (janoszen) [reply pkg/terraform/gather/ovirt/ip.go:121]: After the last update this will cause a panic.

**Review by janoszen — CHANGES_REQUESTED** [review]

File: pkg/terraform/gather/ovirt/ip.go (hunk 3 above)
Comment (eslutsky) [reply pkg/terraform/gather/ovirt/ip.go]: done

File: pkg/terraform/gather/ovirt/ip.go (hunk 4 above)
Comment (eslutsky) [reply pkg/terraform/gather/ovirt/ip.go]: done

File: pkg/terraform/gather/ovirt/ip.go (hunk 5 above)
Comment (eslutsky) [reply pkg/terraform/gather/ovirt/ip.go:125]: done

File: pkg/terraform/gather/ovirt/ip.go (hunk 6 above)
Comment (eslutsky) [reply pkg/terraform/gather/ovirt/ip.go]: done

**Review by janoszen — APPROVED** [review]

**Comment by Gal-Zaidman** [comment]
/assign @staebler 

Required for https://github.com/openshift/installer/pull/5034

**Pull request merged**